Documentation on Sealing Segment (and some minor bug fixes) (#3824)
onflow/flow-go: model/flow/sealing_segment.md

# Sealing Segment

The `SealingSegment` is a section of the finalized chain. It is part of the data need to
initialize a new node to join the network. Informally, the `SealingSegment` is continuous section
of recently finalized blocks that is long enough for the new node to execute its business logic.   

> ## history length covered by the Sealing Segment 

Reviewer: ```suggestion
## History length covered by the Sealing Segment 
```

The `SealingSegment` is created from a `protocol.Snapshot` via the method `SealingSegment`.
Lets denote the block that the `protocol.Snapshot` refers to as `head`. Per convention, 
`head` must be a finalized block. 

### Part 1: from `head` back to the latest sealed block 

> The SealingSegment is the a chain segment such that the last block (greatest height)

Reviewer: ```suggestion
The SealingSegment is the chain segment such that the last block (greatest height)
```

requirement for the history: it shouldn't be longer than necessary. So
extending the chain segment above to `A <- B <- C <- D <- E(SA, SB)` would
be an invalid value for field `SealingSegment.Blocks`.

### Part 2: `ExtraBlocks`

In addition, the `SealingSegment` contains the field `ExtraBlocks`:

```golang
// ExtraBlocks [optional] holds ancestors of `Blocks` in ascending height order. These blocks
// are connecting to `Blocks[0]` (the lowest block of sealing segment). Formally, let `l`
// be the length of `ExtraBlocks`, then ExtraBlocks[l-1] is the _parent_ of `Blocks[0]`.
// These extra blocks are included in order to ensure that a newly bootstrapped node
// knows about all entities which might be referenced by blocks which extend from
// the sealing segment.
// ExtraBlocks are stored separately from Blocks, because not all node roles need
// the same amount of history. Blocks represents the minimal possible required history;
// ExtraBlocks represents additional role-required history.
ExtraBlocks []*Block
```

**In case `head` contains multiple seals, we need _all_ the sealed blocks**, for the following reason:

> * All nodes locally maintain a copy of the protocol state. A system event may change the sate of the protocol state. 

Reviewer: ```suggestion
* All nodes locally maintain a copy of the protocol state. A service event may change the state of the protocol state. 
```

* For Byzantine resilience, we don't want protocol-state changes to take effect immediately. Therefore, we process
  system events only after receiving a QC for the block.
  
  Now let us consider the situation where a newly initialized node comes online and processes the first child of `head`.
  Lets reuse the example from above, where our head was block `E` and we are now processing the child `X`
  ```
            A  <- B <- C <- D <- E(SA, SB) <- X
  ├═══════════┤  ├───────────────────────┤   new
   ExtraBlocks              Blocks          block
  ```
  `X` carries the QC for `E`, hence the protocol-state changes in `E` take effect for `X`. Therefore, when processing `X`,
  we go through the seals in `E` and look through the sealed execution results for service events. 
* As the service events are order-sensitive, we need to process the seals in the correct order, which is by increasing height
  of the sealed block. The seals don't contain the block's height information, hence we need to resolve the block. 

> **Extended history to check for duplicated collection guarantees in blocks** is required by nodes _validate_ block

Reviewer: ```suggestion
**Extended history to check for duplicated collection guarantees in blocks** is required by nodes that _validate_ block
```

payloads (e.g. consensus nodes). Also Access Nodes require these blocks. Collections expire after `flow.DefaultTransactionExpiry` blocks.
Hence, we desire a history of `flow.DefaultTransactionExpiry` blocks. However, there is the edge case of a recent spork (or genesis),
where the history is simply less that `flow.DefaultTransactionExpiry`. 

### Formal definition

The descriptions from the previous section can be formalized as follows

* (i) The highest sealed block as of `head` needs to be included in the sealing segment.
  This is relevant if `head` does not contain any seals.

> * (ii) All blocks that are sealed by `head`. This is relevant if head` contains _multiple_ seals.

Reviewer: ```suggestion
* (ii) All blocks that are sealed by `head`. This is relevant if `head` contains _multiple_ seals.
```
